google-cloud-vision/lib/google/cloud/vision/v1/helpers.rb

```diff
@@ -0,0 +1,971 @@
+          # @param options [Google::Gax::CallOptions]
+          #   Optional. Overrides the default settings for this call, e.g, timeout,
+          #   retries, etc.
+          # @param async [Boolean]
+          #   Optional. Defaults to false. Specifies whether to preform the request synchronously and return a 
+          #   BatchAnnotateImagesResponse instance or to return a Google::Gax::Operation.
+          # @param mime_type [String]
+          #   Required only if async is true.
+          # @param batch_size [Integer]
+          #   Optional. Defaults to 10. When async is true, this specifies the number of input files per output json.
+          # @param destination [String]
+          #   Required only if async is true. A Google Cloud Storage location for storing the output.
+          # @yield [result, operation] Access the result along with the RPC operation
+          # @yieldparam result [Google::Cloud::Vision::V1p3beta1::BatchAnnotateImagesResponse]
+          # @yieldparam operation [GRPC::ActiveCall::Operation]
+          # @return [Google::Cloud::Vision::V1::BatchAnnotateImagesResponse, Google::Gax::Operation]
+          # @example
+          #   require "google/cloud/vision"
+          #
+          #   image_annotator_client = Google::Cloud::Vision::ImageAnnotator.new(version: :v1)
+          #
+          #   response = image_annotator_client.web_detection image: "path\to\image.png"
+          #   response.responses.each do |res|
+          #     puts res
+          #   end
+
+          def web_detection \
+            images: [], 
+            image: nil,
+            max_results: 10,
+            options: nil,
+            async: false,
+            mime_type: nil,
+            batch_size: 10,
+            destination: nil,
+            &blk
+
+            feature = { type: :WEB_DETECTION }
+            feature[:max_results] = max_results
+            images << image if image
+
+            formatted_images = images.map do |img|
+              formatted_image = normalize_image img
+              formatted_image[:mime_type] = mime_type if mime_type
+              formatted_image
+            end
+
+            requests = formatted_images.map do |img|
+              {
+                image: img,
+                features: [feature]
+              }
+            end
+
+            if async
+              requests.map! do |request|
+                {
+                  input_config: request[:image],
+                  features: request[:features],
+                  output_config: {
+                    gcs_destination: destination,
+                    batch_size: batch_size
+                  }
+                }
+              end
+              async_batch_annotate_files requests, options
+            else
+              batch_annotate_images requests, options, &blk
+            end
+          end
+
+          private
+
+          def normalize_image image
+            formatted_image = {}
+            if File.file? image
```

Comment:
Can users pass in a File object?

Reply:
Or an IO object?

Reply:
yes

Reply:
responded before the second comment. Adding IO object handling.

Reply:
A few things here.

(1) I suggest making explicit type checks, because it's really easy to miss cases if you don't. For example, not all IO objects will succeed here. `File.file?` raises a TypeError if you pass it a StringIO. You can check for `respond_to? :binmode` to catch IO and IO-like objects, and use `is_a? String` for strings.
(2) Don't dup an IO object.
(3) You don't need to modify `formatted_image`. Just return the required expression from each of the code paths in the big `if` statement.

So maybe something like:

```ruby
def normalize_image image
  formatted_image =
    if image.respond_to? :binmode
      { content: image.binmode.read }
    elsif image.is_a? String
      if File.file? image
        { content: File.binread image }
      elsif image =~ URI::DEFAULT_PARSER.make_regexp
        if URI(image).scheme == "gs"
          { source: { gcs_image_uri: image } }
        else
          { source: { image_uri: image } }
        end
      end
    end
  raise TypeError.new("Image must be a filepath, url, or IO object") unless formatted_image
  formatted_image
end
```